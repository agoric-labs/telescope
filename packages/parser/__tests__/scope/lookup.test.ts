import { ProtoStore } from '../../src/index'
import { getObjectName, lookupSymbolScopes, lookup, getPackageAndNestedFromStr } from '../../src';
import { getTestProtoStore } from '../../test-utils';
import { getMiscTestProtoStore } from '@cosmology/ast/test-utils';
const store = getTestProtoStore();

const miscStore = getMiscTestProtoStore({
  prototypes: {
    excluded: {
      // hard exclude faulty proto files
      hardProtos: ["google/api/expr/v1alpha1/eval1.proto"],
    },
  },
});
miscStore.traverseAll();

describe('test nested msg', () => {
  const ref = miscStore.findProto('misc/nest.proto');
  it('TestNest_Graph', () => {
      const Graph = lookup(miscStore, ref, 'TestNest_Graph');
      expect(Graph).toMatchSnapshot();
  });
});

// you may need to fix TRAVERSE

describe('double import bug', () => {
    store.traverseAll();
    const ref = store.findProto('ibc/core/types/v1/genesis.proto');
    it('imports/exports', () => {
        expect(ref.traversed.parsedExports).toMatchSnapshot();
        expect(ref.traversed.parsedImports).toMatchSnapshot();
        expect(ref.traversed.importNames).toMatchSnapshot();
    })
    it('GenesisState', () => {
        const GenesisState = lookup(store, ref, 'GenesisState');
        expect(GenesisState).toMatchSnapshot();
        expect(GenesisState.scope).toMatchSnapshot();
        const result1 = getObjectName('GenesisState', GenesisState.scope);
        expect(result1).toEqual('GenesisState')
    });
});

describe('proto import self', () => {
    store.traverseAll();
    const ref = store.findProto('osmosis/gamm/pool-models/balancer/balancerPool.proto');
    it('imports/exports', () => {
        expect(ref.traversed.parsedExports).toMatchSnapshot();
        expect(ref.traversed.parsedImports).toMatchSnapshot();
        expect(ref.traversed.importNames).toMatchSnapshot();
    })
});

describe('proto example', () => {
    store.traverseAll();
    // EXAMPLE: google/logging/v2/logging_metrics
    // EXAMPLE: google/api/servicecontrol/v1/distribution
    const ref = store.findProto('google/logging/v2/logging_metrics.proto');
    it('imports/exports', () => {
        // console.log(ref.traversed.parsedExports);
        // console.log(ref.traversed.parsedImports);
        // console.log(ref.traversed.importNames);
        expect(ref.traversed.parsedExports).toMatchSnapshot();
        expect(ref.traversed.parsedImports).toMatchSnapshot();
        expect(ref.traversed.importNames).toMatchSnapshot();
    })
});

describe('name resolution', () => {
    // LookupSymbol("foo.bar", "baz.moo.corge")
    // https://github.com/protocolbuffers/protobuf/blob/main/src/google/protobuf/descriptor.cc#L3798-L3812
    it('lookup', () => {
        // This first searches siblings of relative_to, then siblings of its
        // parents, etc.

        // For example, LookupSymbol("foo.bar", "baz.moo.corge") makes
        // FindSymbol("baz.moo.foo.bar")
        // FindSymbol("baz.foo.bar"),
        // FindSymbol("foo.bar").

        const lookups = lookupSymbolScopes('foo.bar', 'baz.moo.corge');
        expect(lookups).toEqual(
            ['baz.moo.foo.bar', 'baz.foo.bar', 'foo.bar']
        )
    });
    it('correct Ledger lookup', () => {
        const lookups = lookupSymbolScopes('hd.v1.BIP44Params', 'cosmos.crypto.keyring.v1.Ledger');
        expect(lookups).toEqual(
            [
                'cosmos.crypto.keyring.v1.hd.v1.BIP44Params',
                'cosmos.crypto.keyring.hd.v1.BIP44Params',
                'cosmos.crypto.hd.v1.BIP44Params', // THIS ENDS UP BEING THE WINNER
                'cosmos.hd.v1.BIP44Params',
                'hd.v1.BIP44Params'
            ]
        )
    });
    it('fully qualified name', () => {
        const lookups = lookupSymbolScopes('.hd.v1.BIP44Params', 'cosmos.crypto.keyring.v1.Ledger');
        expect(lookups).toEqual(
            [
                'hd.v1.BIP44Params'
            ]
        )
    });
    it('bad Ledger lookup', () => {
        const lookups = lookupSymbolScopes('v1.BIP44Params', 'cosmos.crypto.keyring.v1.Ledger');
        expect(lookups).toEqual(
            [
                'cosmos.crypto.keyring.v1.v1.BIP44Params',
                'cosmos.crypto.keyring.v1.BIP44Params',
                'cosmos.crypto.v1.BIP44Params',
                'cosmos.v1.BIP44Params',
                'v1.BIP44Params'
            ]
        )
    });
    describe('package matching', () => {
        it('BIP44Params', () => {
            const t = 'cosmos.crypto.hd.v1.BIP44Params';
            const p = 'cosmos.crypto.hd.v1';
            const r = getPackageAndNestedFromStr(t, p);
            expect(r).toEqual({ "nested": "BIP44Params", "package": "cosmos.crypto.hd.v1" })
        })
        it('PoolAssets', () => {
            const t = 'osmosis.gamm.v1beta1.PoolAsset';
            const p = 'osmosis.gamm.v1beta1';
            const r = getPackageAndNestedFromStr(t, p);
            expect(r).toEqual({ "nested": "PoolAsset", "package": "osmosis.gamm.v1beta1" })
        })
        it('google.api.Distribution.BucketOptions', () => {
            const t = 'google.api.Distribution.BucketOptions';
            const p = 'google.api';
            const r = getPackageAndNestedFromStr(t, p);
            expect(r).toEqual({ "nested": "Distribution.BucketOptions", "package": "google.api" })
        })
        it('parseScope', () => {
            expect(store.parseScope(
                'google.api.Distribution.BucketOptions'
            )).toEqual(
                {
                    nested: "Distribution.BucketOptions",
                    package: "google.api"
                }
            );
            expect(store.parseScope(
                'google.api.Distribution'
            )).toEqual(
                {
                    nested: "Distribution",
                    package: "google.api"
                }
            );
            expect(store.parseScope(
                'osmosis.gamm.v1beta1.PoolAsset'
            )).toEqual(
                {
                    nested: "PoolAsset",
                    package: "osmosis.gamm.v1beta1"
                }
            );
            expect(store.parseScope(
                'cosmos.crypto.multisig.v1beta1.CompactBitArray'
            )).toEqual(
                {
                    nested: "CompactBitArray",
                    package: "cosmos.crypto.multisig.v1beta1"
                }
            );
        })
        it('parseScope akash', () => {
            expect(store.parseScope(
                'cosmos.base.v1beta1.Coin'
            )).toEqual(
                {
                    nested: "Coin",
                    package: "cosmos.base.v1beta1"
                }
            );
        })
    })
})
