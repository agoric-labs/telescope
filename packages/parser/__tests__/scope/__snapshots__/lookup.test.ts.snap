// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`double import bug GenesisState 1`] = `
{
  "comment": "GenesisState defines the ibc module's genesis state.",
  "fields": {
    "channelGenesis": {
      "comment": "ICS004 - Channel genesis state",
      "id": 3,
      "import": "ibc/core/channel/v1/genesis.proto",
      "importedName": "ibc.core.channel.v1.GenesisState",
      "isNestedMsg": false,
      "options": {
        "(gogoproto.moretags)": "yaml:channel_genesis",
        "(gogoproto.nullable)": false,
        "(telescope:camel)": "channelGenesis",
        "(telescope:name)": "channelGenesis",
        "(telescope:orig)": "channel_genesis",
      },
      "parsedType": {
        "name": "GenesisState",
        "type": "Type",
      },
      "scope": [
        "ibc.core.channel.v1",
      ],
      "scopeType": "protoImport",
      "type": "ibc.core.channel.v1.GenesisState",
    },
    "clientGenesis": {
      "comment": "ICS002 - Clients genesis state",
      "id": 1,
      "import": "ibc/core/client/v1/genesis.proto",
      "importedName": "ibc.core.client.v1.GenesisState",
      "isNestedMsg": false,
      "options": {
        "(gogoproto.moretags)": "yaml:client_genesis",
        "(gogoproto.nullable)": false,
        "(telescope:camel)": "clientGenesis",
        "(telescope:name)": "clientGenesis",
        "(telescope:orig)": "client_genesis",
      },
      "parsedType": {
        "name": "GenesisState",
        "type": "Type",
      },
      "scope": [
        "ibc.core.client.v1",
      ],
      "scopeType": "protoImport",
      "type": "ibc.core.client.v1.GenesisState",
    },
    "connectionGenesis": {
      "comment": "ICS003 - Connections genesis state",
      "id": 2,
      "import": "ibc/core/connection/v1/genesis.proto",
      "importedName": "ibc.core.connection.v1.GenesisState",
      "isNestedMsg": false,
      "options": {
        "(gogoproto.moretags)": "yaml:connection_genesis",
        "(gogoproto.nullable)": false,
        "(telescope:camel)": "connectionGenesis",
        "(telescope:name)": "connectionGenesis",
        "(telescope:orig)": "connection_genesis",
      },
      "parsedType": {
        "name": "GenesisState",
        "type": "Type",
      },
      "scope": [
        "ibc.core.connection.v1",
      ],
      "scopeType": "protoImport",
      "type": "ibc.core.connection.v1.GenesisState",
    },
  },
  "keyTypes": [],
  "name": "GenesisState",
  "nested": null,
  "oneofs": undefined,
  "options": undefined,
  "package": "ibc.core.types.v1",
  "scope": [
    "ibc.core.types.v1",
  ],
  "type": "Type",
}
`;

exports[`double import bug GenesisState 2`] = `
[
  "ibc.core.types.v1",
]
`;

exports[`double import bug imports/exports 1`] = `
{
  "GenesisState": true,
}
`;

exports[`double import bug imports/exports 2`] = `
{
  "ibc/core/channel/v1/genesis.proto": [
    "GenesisState",
  ],
  "ibc/core/client/v1/genesis.proto": [
    "GenesisState",
  ],
  "ibc/core/connection/v1/genesis.proto": [
    "GenesisState",
  ],
}
`;

exports[`double import bug imports/exports 3`] = `
{
  "ibc/core/channel/v1/genesis.proto": {
    "GenesisState": "GenesisState3",
  },
  "ibc/core/client/v1/genesis.proto": {
    "GenesisState": "GenesisState1",
  },
  "ibc/core/connection/v1/genesis.proto": {
    "GenesisState": "GenesisState2",
  },
}
`;

exports[`proto example imports/exports 1`] = `
{
  "CreateLogMetricRequest": true,
  "DeleteLogMetricRequest": true,
  "GetLogMetricRequest": true,
  "ListLogMetricsRequest": true,
  "ListLogMetricsResponse": true,
  "LogMetric": true,
  "UpdateLogMetricRequest": true,
}
`;

exports[`proto example imports/exports 2`] = `
{
  "google/api/distribution.proto": [
    "Distribution_BucketOptions",
  ],
  "google/api/metric.proto": [
    "MetricDescriptor",
  ],
  "google/protobuf/timestamp.proto": [
    "Timestamp",
  ],
}
`;

exports[`proto example imports/exports 3`] = `
{
  "google/api/distribution.proto": {
    "Distribution_BucketOptions": "Distribution_BucketOptions",
  },
  "google/api/metric.proto": {
    "MetricDescriptor": "MetricDescriptor",
  },
  "google/protobuf/timestamp.proto": {
    "Timestamp": "Timestamp",
  },
}
`;

exports[`proto import self imports/exports 1`] = `
{
  "Pool": true,
  "PoolAsset": true,
  "PoolParams": true,
  "SmoothWeightChangeParams": true,
}
`;

exports[`proto import self imports/exports 2`] = `
{
  "cosmos/base/v1beta1/coin.proto": [
    "Coin",
  ],
  "google/protobuf/duration.proto": [
    "Duration",
  ],
  "google/protobuf/timestamp.proto": [
    "Timestamp",
  ],
}
`;

exports[`proto import self imports/exports 3`] = `
{
  "cosmos/base/v1beta1/coin.proto": {
    "Coin": "Coin",
  },
  "google/protobuf/duration.proto": {
    "Duration": "Duration",
  },
  "google/protobuf/timestamp.proto": {
    "Timestamp": "Timestamp",
  },
}
`;

exports[`test nested msg TestNest_Graph 1`] = `
{
  "comment": null,
  "fields": {
    "name": {
      "comment": null,
      "id": 1,
      "isScalar": true,
      "options": {
        "(telescope:camel)": "name",
        "(telescope:name)": "name",
        "(telescope:orig)": "name",
      },
      "parsedType": {
        "name": "string",
        "type": "native",
      },
      "type": "string",
      "typeNum": 13,
    },
  },
  "keyTypes": [],
  "name": "Graph",
  "nested": null,
  "nestedMsgTypeName": "TestNest_Graph",
  "oneofs": undefined,
  "options": undefined,
  "package": "misc",
  "scope": [
    "misc",
    "TestNest",
  ],
  "type": "Type",
}
`;
